File: src/app/chats/backend/migrar_mysql_a_mongo.js
// migrar_mysql_a_mongo.js
// Migración avanzada de omnicanalidad (MySQL/MariaDB) a MongoDB
// Ejecuta: node migrar_mysql_a_mongo.js
// Requiere: npm install mysql2 mongoose dotenv

require("dotenv").config();
const mysql = require("mysql2/promise");
const mongoose = require("mongoose");
const { ObjectId } = require("mongodb");

// Configuración desde .env o valores por defecto
const MYSQL_CONFIG = {
  host: process.env.MYSQL_HOST || "localhost",
  port: process.env.MYSQL_PORT ? parseInt(process.env.MYSQL_PORT) : 3306,
  user: process.env.MYSQL_USER || "root",
  password: process.env.MYSQL_PASSWORD || "",
  database: process.env.MYSQL_DB || "omnicanalidad",
};
const MONGO_URI = process.env.MONGO_URI || "mongodb://localhost:27017";
const MONGO_DB = process.env.MONGO_DB || "omnicanalidad";

function parseEtiquetas(etiquetaStr, colorStr) {
  if (!etiquetaStr) return [];
  const nombres = etiquetaStr
    .split("|")
    .map((s) => s.trim())
    .filter(Boolean);
  const colores = (colorStr || "").split("|").map((s) => s.trim());
  return nombres.map((nombre, i) => ({ nombre, color: colores[i] || null }));
}

function toDateUTC(dt) {
  if (!dt) return null;
  if (dt instanceof Date) return dt;
  return new Date(dt + "Z");
}

async function main() {
  const mysqlConn = await mysql.createConnection(MYSQL_CONFIG);
  // Conectar solo a la URI base
  const mongoConn = await mongoose.connect(MONGO_URI);
  // Seleccionar la base de datos deseada
  const db = mongoose.connection.useDb(MONGO_DB).db;

  // Utilidad para limpiar y crear índices
  async function resetCollection(name, indexes = []) {
    const col = db.collection(name);
    await col.deleteMany({});
    for (const idx of indexes) {
      await col.createIndex(idx.fields, idx.options || {});
    }
    return col;
  }

  // 1. Etiquetas
  const etiquetasCol = await resetCollection("etiquetas", [
    { fields: { nombre: 1 }, options: { unique: true } },
  ]);
  const [etiquetas] = await mysqlConn.execute("SELECT * FROM etiquetas");
  if (etiquetas.length) {
    await etiquetasCol.insertMany(
      etiquetas.map((r) => ({
        legacyId: r.id,
        nombre: r.nombre,
        color: r.color,
      })),
    );
    console.log(`Migradas ${etiquetas.length} etiquetas.`);
  }

  // 2. EntrantesId
  const entrantesCol = await resetCollection("entrantes", [
    { fields: { numero: 1 }, options: { unique: true } },
  ]);
  const [entrantes] = await mysqlConn.execute("SELECT * FROM entrantesId");
  if (entrantes.length) {
    await entrantesCol.insertMany(
      entrantes.map((r) => ({
        legacyId: r.id,
        mensajeId: r.mensaje_id,
        numero: r.numero,
      })),
    );
    console.log(`Migrados ${entrantes.length} entrantesId.`);
  }

  // 3. Estados de conexión
  const estadosCol = await resetCollection("estados_conexion", [
    { fields: { usuario: 1 }, options: { unique: true } },
  ]);
  const [estados] = await mysqlConn.execute("SELECT * FROM estados_conexion");
  if (estados.length) {
    await estadosCol.insertMany(
      estados.map((r) => ({
        legacyId: r.id,
        usuario: r.usuario,
        estado: r.estado,
        fechaActualizacion: toDateUTC(r.fecha_actualizacion),
      })),
    );
    console.log(`Migrados ${estados.length} estados_conexion.`);
  }

  // 4. Respuestas rápidas
  const respuestasCol = await resetCollection("respuestas_rapidas", [
    { fields: { agenteId: 1, comando: 1 }, options: { unique: true } },
  ]);
  const [respuestas] = await mysqlConn.execute(
    "SELECT * FROM respuestas_rapidas",
  );
  if (respuestas.length) {
    await respuestasCol.insertMany(
      respuestas.map((r) => ({
        legacyId: r.id,
        agenteId: r.agente_id,
        comando: r.comando,
        respuesta: r.respuesta,
        createdAt: toDateUTC(r.created_at),
      })),
    );
    console.log(`Migradas ${respuestas.length} respuestas_rapidas.`);
  }

  // 5. Conversaciones entre agentes
  const convAgenteCol = await resetCollection("conversaciones_agente", [
    { fields: { salaInternaKey: 1 }, options: { unique: true } },
    { fields: { agenteOrigen: 1 } },
    { fields: { agenteDestino: 1 } },
  ]);
  const [convAgente] = await mysqlConn.execute(
    "SELECT * FROM conversaciones_agente",
  );
  const convAgenteIdMap = {};
  if (convAgente.length) {
    const docs = convAgente.map((r) => {
      const oid = new ObjectId();
      convAgenteIdMap[r.id] = oid;
      return {
        _id: oid,
        legacyId: r.id,
        salaInternaKey: r.sala_interna_key,
        agenteOrigen: r.agente_origen_exten,
        agenteDestino: r.agente_destino_exten,
        estado: r.estado,
        inicio: toDateUTC(r.inicio),
        fin: toDateUTC(r.fin),
        ultimaActividad: toDateUTC(r.ultima_actividad),
      };
    });
    await convAgenteCol.insertMany(docs);
    console.log(`Migradas ${docs.length} conversaciones_agente.`);
  }

  // 6. Mensajes internos
  const mensajesInternosCol = await resetCollection("mensajes_internos", [
    { fields: { conversacionId: 1, ts: 1 } },
    { fields: { emisorExten: 1 } },
    { fields: { receptorExten: 1 } },
  ]);
  const [mensajesInternos] = await mysqlConn.execute(
    "SELECT * FROM mensajes_internos",
  );
  if (mensajesInternos.length) {
    await mensajesInternosCol.insertMany(
      mensajesInternos.map((r) => ({
        legacyId: r.id,
        conversacionId: convAgenteIdMap[r.conversacion_id] || r.conversacion_id,
        emisorExten: r.emisor_exten,
        receptorExten: r.receptor_exten,
        mensaje: r.mensaje,
        tipo: r.tipo,
        ts: toDateUTC(r.ts),
        archivoUrl: r.archivo_url,
        leidoEn: toDateUTC(r.leido_en),
      })),
    );
    console.log(`Migrados ${mensajesInternos.length} mensajes_internos.`);
  }

  // 7. Tipificaciones (indexadas por id_conversacion)
  const [tipifs] = await mysqlConn.execute("SELECT * FROM idTipificaciones");
  const tipifMap = {};
  for (const t of tipifs) {
    tipifMap[String(t.id_conversacion)] = {
      texto: t.tipificacion,
      observacion: t.observacion,
    };
  }

  // 8. Conversaciones (clientes)
  const conversacionesCol = await resetCollection("conversaciones", [
    { fields: { legacyId: 1 }, options: { unique: true } },
    { fields: { contactoId: 1 } },
    { fields: { agenteId: 1 } },
    { fields: { estado: 1 } },
    { fields: { origen: 1 } },
    { fields: { inicio: 1 } },
    {
      fields: { agenteId: 1, telefono: 1, salaId: 1 },
      options: { unique: true, sparse: true },
    },
  ]);
  const [convs] = await mysqlConn.execute("SELECT * FROM conversaciones");
  const convIdMap = {};
  if (convs.length) {
    const docs = convs.map((r) => {
      const oid = new ObjectId();
      convIdMap[r.id] = oid;
      const tipif = tipifMap[String(r.id)];
      return {
        _id: oid,
        legacyId: r.id,
        contactoId: r.contacto_id,
        agenteId: r.agente_id,
        estado: r.estado,
        estadoConexion: r.estado_conexion,
        marca: r.marca,
        inicio: toDateUTC(r.inicio),
        fin: toDateUTC(r.fin),
        tipificacion: tipif,
        etiquetas: parseEtiquetas(r.etiqueta, r.color),
        etiqueta2: r.etiqueta_2,
        observaciones: r.observaciones,
        origen: r.origen,
        telefono: r.telefono,
        salaId: r.sala_id,
        cola: r.cola,
      };
    });
    await conversacionesCol.insertMany(docs);
    console.log(`Migradas ${docs.length} conversaciones.`);
  }

  // 9. Mensajes
  const mensajesCol = await resetCollection("mensajes", [
    { fields: { conversacionId: 1, ts: 1 } },
    { fields: { legacyId: 1 } },
  ]);
  const [mensajes] = await mysqlConn.execute("SELECT * FROM mensajes");
  if (mensajes.length) {
    const BATCH = 500;
    let total = 0;
    let batch = [];
    for (const r of mensajes) {
      const legacyConv = r.conversacion_id;
      const convOid = convIdMap[legacyConv] || legacyConv;
      batch.push({
        legacyId: r.id,
        conversacionId: convOid,
        emisor: r.emisor,
        mensaje: r.mensaje,
        tipo: r.tipo,
        ts: toDateUTC(r.ts),
        archivoUrl: r.archivo_url,
        origen: r.origen,
      });
      if (batch.length >= BATCH) {
        await mensajesCol.insertMany(batch);
        total += batch.length;
        batch = [];
      }
    }
    if (batch.length) {
      await mensajesCol.insertMany(batch);
      total += batch.length;
    }
    console.log(`Migrados ${total} mensajes.`);
  }

  await mysqlConn.end();
  await mongoose.disconnect();
  console.log("\n✅ Migración completada exitosamente.");
}

main().catch((err) => {
  console.error("Error en la migración:", err);
  process.exit(1);
});

File: chats/vue-chats/backend/migrar_mysql_a_mongo.py
"""
Migración: omnicanalidad (MariaDB/MySQL) → MongoDB
Requiere:
  pip install pymysql pymongo python-dotenv

Variables de entorno (.env):
  MYSQL_HOST, MYSQL_PORT, MYSQL_USER, MYSQL_PASSWORD, MYSQL_DB
  MONGO_URI, MONGO_DB
"""

import os
import re
from datetime import timezone
from dotenv import load_dotenv
import pymysql
import pymysql.cursors
from pymongo import MongoClient, UpdateOne, ASCENDING
from bson import ObjectId

load_dotenv()

def get_mysql():
    return pymysql.connect(
        host=os.getenv("MYSQL_HOST", "localhost"),
        port=int(os.getenv("MYSQL_PORT", 3306)),
        user=os.getenv("MYSQL_USER", "root"),
        password=os.getenv("MYSQL_PASSWORD", ""),
        database=os.getenv("MYSQL_DB", "omnicanalidad"),
        cursorclass=pymysql.cursors.DictCursor,
        charset="utf8mb4",
    )

def get_mongo():
    client = MongoClient(os.getenv("MONGO_URI", "mongodb://localhost:27017"))
    return client[os.getenv("MONGO_DB", "omnicanalidad")]

def parse_etiquetas(etiqueta_str, color_str):
    if not etiqueta_str:
        return []
    nombres = [n.strip() for n in etiqueta_str.split("|") if n.strip()]
    colores = [c.strip() for c in (color_str or "").split("|") if c.strip()]
    return [
        {"nombre": n, "color": colores[i] if i < len(colores) else None}
        for i, n in enumerate(nombres)
    ]

def to_utc(dt):
    if dt is None:
        return None
    if dt.tzinfo is None:
        return dt.replace(tzinfo=timezone.utc)
    return dt

def migrate_etiquetas(mysql_conn, mongo_db):
    print("→ Migrando etiquetas...")
    col = mongo_db["etiquetas"]
    col.drop()
    col.create_index([("nombre", ASCENDING)], unique=True)
    with mysql_conn.cursor() as cur:
        cur.execute("SELECT * FROM etiquetas")
        rows = cur.fetchall()
    docs = [
        {
            "legacyId": r["id"],
            "nombre": r["nombre"],
            "color": r["color"],
        }
        for r in rows
    ]
    if docs:
        col.insert_many(docs)
    print(f"   {len(docs)} etiquetas insertadas.")

def migrate_entrantes(mysql_conn, mongo_db):
    print("→ Migrando entrantesId...")
    col = mongo_db["entrantes"]
    col.drop()
    col.create_index([("numero", ASCENDING)], unique=True)
    with mysql_conn.cursor() as cur:
        cur.execute("SELECT * FROM entrantesId")
        rows = cur.fetchall()
    docs = [
        {
            "legacyId": r["id"],
            "mensajeId": r["mensaje_id"],
            "numero": r["numero"],
        }
        for r in rows
    ]
    if docs:
        col.insert_many(docs)
    print(f"   {len(docs)} entrantesId insertados.")

def migrate_estados_conexion(mysql_conn, mongo_db):
    print("→ Migrando estados_conexion...")
    col = mongo_db["estados_conexion"]
    col.drop()
    col.create_index([("usuario", ASCENDING)], unique=True)
    with mysql_conn.cursor() as cur:
        cur.execute("SELECT * FROM estados_conexion")
        rows = cur.fetchall()
    docs = [
        {
            "legacyId": r["id"],
            "usuario": r["usuario"],
            "estado": r["estado"],
            "fechaActualizacion": to_utc(r["fecha_actualizacion"]),
        }
        for r in rows
    ]
    if docs:
        col.insert_many(docs)
    print(f"   {len(docs)} estados insertados.")

def migrate_respuestas_rapidas(mysql_conn, mongo_db):
    print("→ Migrando respuestas_rapidas...")
    col = mongo_db["respuestas_rapidas"]
    col.drop()
    col.create_index([("agenteId", ASCENDING), ("comando", ASCENDING)], unique=True)
    with mysql_conn.cursor() as cur:
        cur.execute("SELECT * FROM respuestas_rapidas")
        rows = cur.fetchall()
    docs = [
        {
            "legacyId": r["id"],
            "agenteId": r["agente_id"],
            "comando": r["comando"],
            "respuesta": r["respuesta"],
            "createdAt": to_utc(r["created_at"]),
        }
        for r in rows
    ]
    if docs:
        col.insert_many(docs)
    print(f"   {len(docs)} respuestas rápidas insertadas.")

def migrate_conversaciones_agente(mysql_conn, mongo_db):
    print("→ Migrando conversaciones_agente...")
    col = mongo_db["conversaciones_agente"]
    col.drop()
    col.create_index([("salaInternaKey", ASCENDING)], unique=True)
    col.create_index([("agenteOrigen", ASCENDING)])
    col.create_index([("agenteDestino", ASCENDING)])
    with mysql_conn.cursor() as cur:
        cur.execute("SELECT * FROM conversaciones_agente")
        rows = cur.fetchall()
    id_map = {}
    docs = []
    for r in rows:
        oid = ObjectId()
        id_map[r["id"]] = oid
        docs.append({
            "_id": oid,
            "legacyId": r["id"],
            "salaInternaKey": r["sala_interna_key"],
            "agenteOrigen": r["agente_origen_exten"],
            "agenteDestino": r["agente_destino_exten"],
            "estado": r["estado"],
            "inicio": to_utc(r["inicio"]),
            "fin": to_utc(r["fin"]),
            "ultimaActividad": to_utc(r["ultima_actividad"]),
        })
    if docs:
        col.insert_many(docs)
    print(f"   {len(docs)} conversaciones_agente insertadas.")
    return id_map

def migrate_mensajes_internos(mysql_conn, mongo_db, conv_agente_id_map):
    print("→ Migrando mensajes_internos...")
    col = mongo_db["mensajes_internos"]
    col.drop()
    col.create_index([("conversacionId", ASCENDING), ("ts", ASCENDING)])
    col.create_index([("emisorExten", ASCENDING)])
    col.create_index([("receptorExten", ASCENDING)])
    with mysql_conn.cursor() as cur:
        cur.execute("SELECT * FROM mensajes_internos")
        rows = cur.fetchall()
    docs = [
        {
            "legacyId": r["id"],
            "conversacionId": conv_agente_id_map.get(r["conversacion_id"], r["conversacion_id"]),
            "emisorExten": r["emisor_exten"],
            "receptorExten": r["receptor_exten"],
            "mensaje": r["mensaje"],
            "tipo": r["tipo"],
            "ts": to_utc(r["ts"]),
            "archivoUrl": r["archivo_url"],
            "leidoEn": to_utc(r["leido_en"]),
        }
        for r in rows
    ]
    if docs:
        col.insert_many(docs)
    print(f"   {len(docs)} mensajes_internos insertados.")

def migrate_conversaciones(mysql_conn, mongo_db):
    print("→ Migrando conversaciones...")
    with mysql_conn.cursor() as cur:
        cur.execute("SELECT * FROM idTipificaciones")
        tipifs_raw = cur.fetchall()
    tipif_map = {
        str(t["id_conversacion"]): {
            "texto": t["tipificacion"],
            "observacion": t["observacion"],
        }
        for t in tipifs_raw
    }
    col = mongo_db["conversaciones"]
    col.drop()
    col.create_index([("legacyId", ASCENDING)], unique=True)
    col.create_index([("contactoId", ASCENDING)])
    col.create_index([("agenteId", ASCENDING)])
    col.create_index([("estado", ASCENDING)])
    col.create_index([("origen", ASCENDING)])
    col.create_index([("inicio", ASCENDING)])
    col.create_index([("agenteId", ASCENDING), ("telefono", ASCENDING), ("salaId", ASCENDING)], unique=True, sparse=True)
    with mysql_conn.cursor() as cur:
        cur.execute("SELECT * FROM conversaciones")
        rows = cur.fetchall()
    id_map = {}
    docs = []
    for r in rows:
        oid = ObjectId()
        id_map[r["id"]] = oid
        key = str(r["id"])
        tipif = tipif_map.get(key)
        docs.append({
            "_id": oid,
            "legacyId": r["id"],
            "contactoId": r["contacto_id"],
            "agenteId": r["agente_id"],
            "estado": r["estado"],
            "estadoConexion": r["estado_conexion"],
            "marca": r["marca"],
            "inicio": to_utc(r["inicio"]),
            "fin": to_utc(r["fin"]),
            "tipificacion": tipif,
            "etiquetas": parse_etiquetas(r["etiqueta"], r["color"]),
            "etiqueta2": r["etiqueta_2"],
            "observaciones": r["observaciones"],
            "origen": r["origen"],
            "telefono": r["telefono"],
            "salaId": r["sala_id"],
            "cola": r["cola"],
        })
    if docs:
        col.insert_many(docs)
    print(f"   {len(docs)} conversaciones insertadas.")
    return id_map

def migrate_mensajes(mysql_conn, mongo_db, conv_id_map):
    print("→ Migrando mensajes...")
    col = mongo_db["mensajes"]
    col.drop()
    col.create_index([("conversacionId", ASCENDING), ("ts", ASCENDING)])
    col.create_index([("legacyId", ASCENDING)])
    with mysql_conn.cursor() as cur:
        cur.execute("SELECT * FROM mensajes")
        rows = cur.fetchall()
    BATCH = 500
    total = 0
    batch = []
    for r in rows:
        legacy_conv = r["conversacion_id"]
        conv_oid = conv_id_map.get(legacy_conv, legacy_conv)
        batch.append({
            "legacyId": r["id"],
            "conversacionId": conv_oid,
            "emisor": r["emisor"],
            "mensaje": r["mensaje"],
            "tipo": r["tipo"],
            "ts": to_utc(r["ts"]),
            "archivoUrl": r["archivo_url"],
            "origen": r["origen"],
        })
        if len(batch) >= BATCH:
            col.insert_many(batch)
            total += len(batch)
            batch = []
    if batch:
        col.insert_many(batch)
        total += len(batch)
    print(f"   {total} mensajes insertados.")

def main():
    print("=== Migración MySQL → MongoDB: omnicanalidad ===\n")
    mysql = get_mysql()
    mongo = get_mongo()
    try:
        migrate_etiquetas(mysql, mongo)
        migrate_entrantes(mysql, mongo)
        migrate_estados_conexion(mysql, mongo)
        migrate_respuestas_rapidas(mysql, mongo)
        conv_agente_map = migrate_conversaciones_agente(mysql, mongo)
        migrate_mensajes_internos(mysql, mongo, conv_agente_map)
        conv_map = migrate_conversaciones(mysql, mongo)
        migrate_mensajes(mysql, mongo, conv_map)
        print("\n✅ Migración completada exitosamente.")
        print(f"   Base MongoDB: {mongo.name}")
        print(f"   Colecciones: {mongo.list_collection_names()}")
    finally:
        mysql.close()

if __name__ == "__main__":
    main()
